core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -460,25 +493,213 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
         dagViz ++
         maybeExpandDagViz ++
         showAdditionalMetrics ++
+        makeTimeline(stageData.taskData.values.toSeq, currentTime) ++
         <h4>Summary Metrics for {numCompleted} Completed Tasks</h4> ++
         <div>{summaryTable.getOrElse("No tasks have reported metrics yet.")}</div> ++
         <h4>Aggregated Metrics by Executor</h4> ++ executorTable.toNodeSeq ++
         maybeAccumulableTable ++
         <h4>Tasks</h4> ++ taskTable
-
       UIUtils.headerSparkPage(stageHeader, content, parent, showVisualization = true)
     }
   }
 
+  def makeTimeline(tasks: Seq[TaskUIData], currentTime: Long): Seq[Node] = {
+    val executorsSet = new HashSet[(String, String)]
+    var minLaunchTime = Long.MaxValue
+    var maxFinishTime = Long.MinValue
+    var numEffectiveTasks = 0
+
+    val executorsArrayStr = tasks.filter { taskUIData =>
+      val taskInfo = taskUIData.taskInfo
+      taskInfo.successful || taskInfo.running || taskInfo.failed
+    }.sortWith { (uiDataL, uiDataR) =>
+      val taskInfoL = uiDataL.taskInfo
+      val launchTimeL = taskInfoL.launchTime
+
+      val taskInfoR = uiDataR.taskInfo
+      val launchTimeR = taskInfoR.launchTime
+
+      launchTimeL > launchTimeR
+    }.take(MAX_TIMELINE_TASKS).map { taskUIData =>
+      val taskInfo = taskUIData.taskInfo
+      val executorId = taskInfo.executorId
+      val host = taskInfo.host
+      executorsSet += ((executorId, host))
+
+      val taskId = taskInfo.taskId
+      val taskIdWithIndexAndAttempt = s"Task ${taskId}(${taskInfo.id})"
```

Comment:
Could you change this to s"Task ${taskInfo.index} (attempt ${taskInfo.attempt}"? I think that would be more intuitive